busbus/web.py

```diff
@@ -0,0 +1,154 @@
+
+
+EXPAND_TYPES = {
+    'providers': ProviderBase,
+    'agencies': busbus.Agency,
+    'stops': busbus.Stop,
+    'routes': busbus.Route,
+    'arrivals': busbus.Arrival,
+}
+
+
+def unexpand_init(result, to_expand):
+    return ({attr: unexpand(value, to_expand)
+             for attr, value in dict(obj).items()}
+            for obj in result)
+
+
+def unexpand(obj, to_expand):
+    for name, cls in EXPAND_TYPES.items():
+        if isinstance(obj, cls):
+            if name not in to_expand:
+                return {'id': obj.id}
+            else:
+                return {attr: unexpand(value, to_expand)
+                        for attr, value in dict(obj).items()}
+    if isinstance(obj, dict):
+        return {attr: unexpand(value, to_expand)
+                for attr, value in obj.items()}
+    if isinstance(obj, (list, tuple, collections.Iterator)):
+        return (unexpand(value, to_expand) for value in obj)
+    return obj
+
+
+class APIError(Exception):
+
+    def __init__(self, msg, error_code=500):
+        self.msg = msg
+        self.error_code = error_code
+
+
+class Engine(busbus.Engine):
+
+    def __init__(self, *args, **kwargs):
+        # perhaps fix this to use a decorator somehow?
+        self._entity_actions = {
+            ('stops', 'find'): (self.stops_find, 'stops'),
+            ('routes', 'directions'): (self.routes_directions, 'directions'),
+        }
+        super(Engine, self).__init__(*args, **kwargs)
+
+    @cherrypy.popargs('entity', 'action')
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def default(self, entity=None, action=None, **kwargs):
+        if entity is None:
+            return self.help()
+
+        response = {
+            'request': {
+                'status': 'ok',
+                'entity': entity,
+                'params': kwargs,
+            }
+        }
+        to_expand = (kwargs.pop('_expand').split(',')
+                     if '_expand' in kwargs else [])
+        if to_expand:
+            response['request']['expand'] = to_expand
+        if action:
+            response['request']['action'] = action
+            if (entity, action) in self._entity_actions:
+                func, entity = self._entity_actions[(entity, action)]
+                try:
+                    response[entity] = unexpand_init(func(**kwargs), to_expand)
+                except APIError as exc:
+                    response['request']['status'] = 'error'
+                    response['error'] = exc.msg
+                    cherrypy.response.status = exc.error_code
+                return response
+        else:
+            entity_func = getattr(super(Engine, self), entity, None)
+            if entity_func is not None:
+                if '_limit' in kwargs:
+                    limit = int(kwargs.pop('_limit'))
+                    response['request']['limit'] = limit
+                result = getattr(super(Engine, self), entity).where(**kwargs)
```

Comment:
For when a limit is given, is this doing unnecessary work? I mean if I want 1 arrival, is it still going to do the work for all the arrivals and then slice it down to one a couple lines down?